apstra/api_custom_collector.go

```diff
@@ -0,0 +1,257 @@
+	ServiceName             string
+	Platform                CollectorPlatform `json:"platform"`
+	SourceType              string            `json:"source_type"`
+	Cli                     string            `json:"cli"`
+	Query                   Query             `json:"query"`
+	RelaxedSchemaValidation bool              `json:"relaxed_schema_validation"`
+}
+
+// GetAllCollectors gets all the Collectors for all services
+func (o *Client) GetAllCollectors(ctx context.Context) ([]Collector, error) {
+	var response struct {
+		Items map[string]interface{} `json:"items"`
+	}
+	var collectors []Collector
+	// First Reach out to /collectors , we are interested really only in the keys, so that we can pull the collectors
+	err := o.talkToApstra(ctx, &talkToApstraIn{
+		method:      http.MethodGet,
+		urlStr:      fmt.Sprintf(apiUrlCollectors),
+		apiResponse: &response,
+	})
+	if err != nil {
+		return nil, convertTtaeToAceWherePossible(err)
+	}
+
+	for k := range response.Items {
+		cs, err := o.GetCollectorsByServiceName(ctx, k)
+		if err != nil {
+			return nil, err
+		}
+		for _, v := range cs {
+			v.ServiceName = k
+			collectors = append(collectors, v)
+		}
+	}
+	return collectors, nil
+}
+
+// GetCollectorsByServiceName gets all the Collectors that correspond to a particular service
+func (o *Client) GetCollectorsByServiceName(ctx context.Context, name string) ([]Collector, error) {
+	var ace ClientErr
+	var Response struct {
+		Items []Collector `json:"items"`
+	}
+	err := o.talkToApstra(ctx, &talkToApstraIn{
+		method:      http.MethodGet,
+		urlStr:      fmt.Sprintf(apiUrlCollectorsByServiceName, name),
+		apiResponse: &Response,
+	})
+
+	if err != nil {
+		err = convertTtaeToAceWherePossible(err)
+		if errors.As(err, &ace) && ace.Type() == ErrNotfound {
+			return Response.Items, nil
+		}
+		return nil, err
+	}
+
+	for i := range Response.Items {
+		Response.Items[i].ServiceName = name
+	}
+	return Response.Items, nil
+}
+
+// CreateCollector creates a collector
+func (o *Client) CreateCollector(ctx context.Context, in *Collector) error {
+	// Check if this is the first collector for that service name
+	cs, err := o.GetCollectorsByServiceName(ctx, in.ServiceName)
+	if err != nil {
+		return err
+	}
+	var Request struct {
+		ServiceName string      `json:"service_name"`
+		Items       []Collector `json:"collectors"`
+	}
+	Request.ServiceName = in.ServiceName
+	Request.Items = append(Request.Items, *in)
+	// This is the first collector for this service name
+	// So we POST
+	if len(cs) == 0 {
+		err = o.talkToApstra(ctx, &talkToApstraIn{
+			method:   http.MethodPost,
+			urlStr:   fmt.Sprintf(apiUrlCollectors),
+			apiInput: &Request,
+		})
+		return err
+	}
+
+	// There are other collectors, so this is a patch
+	err = o.talkToApstra(ctx, &talkToApstraIn{
+		method:   http.MethodPatch,
+		urlStr:   fmt.Sprintf(apiUrlCollectorsByServiceName, in.ServiceName),
+		apiInput: &Request,
+	})
+	return err
+}
+
+// Update Collector Updates a collector
```

Comment:
Pls make this satisfy the Go Doc comment standard (or remove it)

Reply:
Fixed